apps/voting/contracts/Voting.sol

```diff
@@ -213,6 +240,72 @@ contract Voting is IForwarder, AragonApp {
         _executeVote(_voteId);
     }
 
+    /**
+    * @notice Assign `_delegate` as the delegate for the sender
+    * @param _delegate address to delegate to
+    */
+    function setDelegate(address _delegate) external {
+        require(_delegate != address(0), ERROR_ZERO_ADDRESS_PASSED);
+        require(_delegate != msg.sender, ERROR_SELF_DELEGATE);
+
+        address prevDelegate = delegates[msg.sender].delegate;
+        require(_delegate != prevDelegate, ERROR_DELEGATE_SAME_AS_PREV);
+
+        uint256 votingPower = token.balanceOfAt(msg.sender, getBlockNumber64() - 1);
+        require(votingPower > 0, ERROR_NO_VOTING_POWER);
+
+        if (prevDelegate != address(0)) {
+            _removeDelegatedAddressFor(prevDelegate, msg.sender);
+        }
+        _addDelegatedAddressFor(_delegate, msg.sender);
+    }
+
+    /**
+    * @notice Unassign `_delegate` from the sender
+    */
+    function resetDelegate() external {
+        address prevDelegate = delegates[msg.sender].delegate;
+        require(prevDelegate != address(0), ERROR_DELEGATE_NOT_SET);
+
+        _removeDelegatedAddressFor(prevDelegate, msg.sender);
+        delete delegates[msg.sender];
+    }
+
+    /**
+     * @notice Vote `_supports ? 'yes' : 'no'` in vote #`_voteId` on behalf of the `_voters` list
+     * @param _voteId Id for vote
+     * @param _supports Whether the delegate supports the vote
+     * @param _voters list of voters
+     */
+    function attemptVoteForMultiple(uint256 _voteId, bool _supports, address[] _voters) public voteExists(_voteId) {
+        Vote storage vote_ = votes[_voteId];
+        require(_isValidPhaseToVote(vote_, _supports), ERROR_CAN_NOT_VOTE);
+        bool hasManagedToVote = false;
+
+        for (uint256 i = 0; i < _voters.length; ++i) {
+            address voter = _voters[i];
+            require(_hasVotingPower(vote_, voter), ERROR_NO_VOTING_POWER);
+            if (_canVoteFor(vote_, msg.sender, voter)) {
+                _vote(_voteId, _supports, voter, true);
+                hasManagedToVote = true;
+            }
+        }
+
+        require(hasManagedToVote, ERROR_CAN_NOT_VOTE_FOR);
```

Comment:
suggest to consider implementing some kind of `filter` method to clean the list of ineligible voters (i.e. check power/canVote) and get the list of "eligible" voters, and only then call the vote method in a loop without additional checks.

This will get rid of the logical flag (it is enough to check the length of the filtered list), and, moreover, it will be possible to remove the  event CastVoteAsDelegate emit in the loop. If make it for the list of addresses, it is enough to emit in once after the loop.

e.g. `event CastVoteAsDelegate(uint256 indexed voteId, address indexed delegate, address[] voters)`

pretty sure the other arguments in this event are redundant and generally useless as they are present in other events

Reply:
Addressed in 32a3dd239c42b642c2cbba3065bbf66e2f4a61b0. 

(comment TBA)